### DISCUSSION TITLE
Arduino UNO R4 Support

### DISCUSSION THREAD

**Discussion by cavonlee** [Arduino-IRremote/Arduino-IRremote#1147] [Ideas]
As Arduino UNO R4 is now announced, I think there should be  a support for it.

**Comment by eMUQI** [comment]
Same demand

Hi there,

I've been using the Arduino-IRremote library and I recently got my hands on the new Arduino Uno R4 development board. However, I noticed that the library doesn't currently support this board. Can we get support for the Uno R4 added to the library?

I think this would be a great addition to the library and would benefit a lot of users, especially those looking to use the IR receiver and transmitter modules on the Uno R4.

Thank you for all your hard work on this library and considering this request.

Best regards

**Comment by DIYables** [comment]
I have the same question and demand.

I developer the [DIYables_IRcontroller](https://github.com/DIYables/DIYables_IRcontroller) library based on this library for some specific IR controller. I am looking forward for Uno R4 support.

**Comment by opticaldrive** [comment]
Just curious but could someone try using IRremote on the R4 minima and WiFi? It might work………
Or not…..
Worth a try?

**Reply by eMUQI** [reply]
I tried on UNO R4 minima and got the following results.
```
Arduino\libraries\IRremote\src/private/IRTimer.hpp:1874:2: error: #error Internal code configuration error, no timer functions implemented for this CPU / board
 #error Internal code configuration error, no timer functions implemented for this CPU / board
  ^~~~~

exit status 1

Compilation error: exit status 1
```

You can also check the test results in the UNO R4 Library Compatibility repository maintained by Arduino official.
https://github.com/arduino/uno-r4-library-compatibility#sensors

**Reply by ArminJo** [reply]
Sorry, but I have no Uno R4 at hand, so I cannot adapt the library to it 😢 .

**Reply by opticaldrive** [reply]
That’s sad :(

**Comment by haydenmc** [comment]
Hey friends, I was able to get the library working with my Arduino Uno R4 WiFi, my changes are here https://github.com/haydenmc/Arduino-IRremote/commit/1f05bf6b739b806b0c2562dff71e8663b07f1f40 though looks like I was beat to the punch by a few hours by this change https://github.com/Arduino-IRremote/Arduino-IRremote/commit/c2e55e9d250af25e96ed9f7eb46bfa14b08b1e54 that's already available in master!

FYI that without another fix in the Renesas core https://github.com/arduino/ArduinoCore-renesas/pull/52 you may hit a snag where the IR receive code hits an infinite loop due to this bug https://github.com/arduino/ArduinoCore-renesas/issues/49.

With both of the changes above in place I'm able to successfully receive and transmit IR codes from my Arduino Uno R4 WiFi.

**Comment by ArminJo** [comment]
@haydenmc Great work! I changed my implementation after I saw yours 😀 .
Could you do me a favor and test the latest IRTimer.hpp from this repository?
Thanks!

**Reply by haydenmc** [reply]
Thank you!

I've pulled down the latest `master` and I'm running into a couple issues.

One is that I believe the modification to `PinDefinitionsAndMore.h` is meant to contain an `#elif` instead of an `#if` on line 156, without that change I see this preprocessor error:

```
In file included from C:\Users\Hayden\Source\Arduino-IRremote\examples\ReceiveDump\ReceiveDump.ino:36:0:
C:\Users\Hayden\Source\Arduino-IRremote\examples\ReceiveDump\PinDefinitionsAndMore.h:51:0: error: unterminated #if
 #if defined(__AVR__)
```

The second is that while I am able to successfully send IR commands, receiving them via the ReceiveDump example has stopped working. I'm investigating to see if I can determine the root cause.

EDIT: Looks like things are broken even with my original change. Something fishy is going on 😅

EDIT²: Just kidding, I wasn't properly applying my original commit. Receiving works with my change but is broken with the new one. Doing some more investigating...

**Reply by haydenmc** [reply]
In `IRTimer.hpp` you are missing `s50usTimer.open()` on line 1128. With this change I am able to receive successfully!

**Comment by ArminJo** [comment]
Thanks a lot!
I fixed it now.

**Comment by DIYables** [comment]
Dear @ArminJo 

Thank you so much for fixing it.

Could you please release new version to make it available on Arduino IDE Library Manager ?

**Reply by ArminJo** [reply]
In general, I wait a few days, just to be shure...
